source/opt/licm_pass.cpp

```diff
@@ -0,0 +1,138 @@
+    modified |= ProcessFunction(&f);
+  }
+  return modified;
+}
+
+bool LICMPass::ProcessFunction(ir::Function* f) {
+  bool modified = false;
+  ir::LoopDescriptor loop_descriptor{f};
+
+  // Process each loop in the function
+  for (ir::Loop& loop : loop_descriptor) {
+    modified |= ProcessLoop(&loop, f);
+  }
+  return modified;
+}
+
+bool LICMPass::ProcessLoop(ir::Loop* loop, ir::Function* f) {
+  // Process all nested loops first
+  for (ir::Loop* nested_loop : *loop) {
+    ProcessLoop(nested_loop, f);
+  }
+
+  std::queue<ir::Instruction*> loop_iv_instr{};
+  GatherAllImmediatelyInvariantInstructions(loop, &loop_iv_instr);
+
+  return ProcessInstructionList(loop, &loop_iv_instr);
+}
+
+void LICMPass::GatherAllImmediatelyInvariantInstructions(
+    ir::Loop* loop, std::queue<ir::Instruction*>* loop_iv_instr) {
+  // For each instruction in each basic block in the loop, check if it is
+  // invariant. If so, push it to our invariants queue.
+  for (uint32_t bb_id : loop->GetBlocks()) {
+    ir::BasicBlock* bb = ir_context->get_instr_block(bb_id);
+    for (ir::Instruction& inst : *bb) {
+      if (inst.opcode() != SpvOpPhi && !inst.HasSideEffects() &&
+          AllOperandsOutsideLoop(loop, &inst)) {
+        loop_iv_instr->push(&inst);
+      }
+    }
+  }
+}
+
+void LICMPass::HoistInstruction(ir::BasicBlock* pre_header_bb,
+                                ir::Instruction* inst) {
+  inst->InsertBefore(std::move(&(*pre_header_bb->tail())));
+  ir_context->set_instr_block(inst, pre_header_bb);
+}
+
+bool LICMPass::AllOperandsOutsideLoop(ir::Loop* loop, ir::Instruction* inst) {
+  analysis::DefUseManager* def_use_mgr = ir_context->get_def_use_mgr();
+  bool all_outside_loop = true;
+
+  const std::function<void(uint32_t*)> operand_test_lambda =
+      [&def_use_mgr, &loop, &all_outside_loop](uint32_t* id) {
+        if (loop->IsInsideLoop(def_use_mgr->GetDef(*id))) {
+          all_outside_loop = false;
+        }
+      };
+
+  inst->ForEachInId(operand_test_lambda);
+  return all_outside_loop;
+}
+
+bool LICMPass::ProcessInstructionList(
+    ir::Loop* loop, std::queue<ir::Instruction*>* loop_iv_instr) {
+  if (loop_iv_instr->empty()) {
+    return false;
+  }
+
+  ir::BasicBlock* pre_header_bb = loop->GetPreHeaderBlock();
+
+  // If the input instruction is invariant, push it onto the invariants queue
+  const std::function<void(ir::Instruction*)> check_users_now_invariant =
+      [this, &loop, &loop_iv_instr](ir::Instruction* inst) {
+        if (inst->opcode() != SpvOpPhi && !inst->HasSideEffects() &&
+            AllOperandsOutsideLoop(loop, inst)) {
+          loop_iv_instr->push(inst);
+        }
+      };
+
+  // While there are invariant instructions in the queue, hoist them outside of
+  // the loop, then add their users to the queue if they have become invariant
+  while (!loop_iv_instr->empty()) {
+    ir::Instruction* inst = loop_iv_instr->front();
+    loop_iv_instr->pop();
+    HoistInstruction(pre_header_bb, inst);
```

Comment:
As I think about this, we might be able to simplify this code, and remove the dependence on the def-use chains.  In `GatherAllImmediatelyInvariantInstructions`, if you traverse the blocks in dominance order starting with the header, and you hoist the instruction right away instead of waiting, then that first pass should be all that you need.  No need for the work list.

The idea is that if an instruction is loop invariant, then its input are outside the loop, or they are inside the loop, must dominate the use, and are invariant themselves.  So the feeders will have been moved first.

Personally, I don't like using the def-use chain (use-def chains are fine) because they are expensive to keep up to date.  Eventually we need to split the def-use and use-def chains into two separate analyses, and use the def-use chain rarely.